src/runtime/contrib/dnnl/dnnl_json_runtime.cc

```diff
@@ -269,24 +309,126 @@ class DNNLJSONRuntime : public JSONRuntimeBase {
-        OC = weight_shape[0];              // output channels
+    // Check layout.
+    if (layout_dict.find(data_layout) == layout_dict.end() ||
+        layout_dict.find(kernel_layout) == layout_dict.end()) {
+      LOG(FATAL) << "Unsupported layout: " << data_layout << " " << kernel_layout;
+    }
 
     // Memory shapes.
-    dnnl::memory::dims data_dims = {B, IC};
-    dnnl::memory::dims weight_dims = {OC, IC};
+    dnnl::memory::dims src_dims = input_shape;       // {N, IC, ID, IH, IW}
+    dnnl::memory::dims weights_dims = weight_shape;  // {OC, IC, KD, KH, KW}
+
+    // Check weight shape, transform to `OIHW`
+    if (weights_dims[0] == src_dims[1] && weights_dims[1] == channels) {
+      std::swap(weights_dims[0], weights_dims[1]);
+    }
+    if (kernel_layout == "OIDHW") {
+      kernel_layout = "IODHW";
+    }
+    if (groups > 1) {
+      weights_dims = {groups, 1, input_shape[1] / groups};
+      weights_dims.insert(weights_dims.end(), weight_shape.begin() + 2, weight_shape.end());
+      kernel_layout.insert(0, "G");
+    }
+    dnnl::memory::dims bias_dims = {channels};
+    dnnl::memory::dims dst_dims = out_shape;  // {N, OC, OD, OH, OW}
+    dnnl::memory::dims strides_dims = TransformStr2Dims(str_strides, "strides");
+    dnnl::memory::dims dilates_dims = TransformStr2Dims(str_dilates, "dilates");
+    dnnl::memory::dims padding_dims_l = TransformStr2Dims(str_padding_l, "padding");
+    dnnl::memory::dims padding_dims_r = TransformStr2Dims(str_padding_r, "padding");
+
+    // Memory descriptions.
+    auto deconv_src_md = dnnl::memory::desc(src_dims, dt::f32, layout_dict[data_layout]);
+    auto deconv_weights_md = dnnl::memory::desc(weights_dims, dt::f32, layout_dict[kernel_layout]);
+    auto deconv_bias_md = dnnl::memory::desc(bias_dims, dt::f32, tag::any);
+    auto deconv_dst_md = dnnl::memory::desc(dst_dims, dt::f32, layout_dict[data_layout]);
+
+    // Transposed covn2d description.
+    auto deconv_desc = dnnl::deconvolution_forward::desc(
+        dnnl::prop_kind::forward_inference, dnnl::algorithm::deconvolution_direct, deconv_src_md,
+        deconv_weights_md, deconv_bias_md, deconv_dst_md, strides_dims, dilates_dims,
+        padding_dims_l, padding_dims_r);
+
+    // Enable elementwise post-ops.
+    auto deconv2d_prim_desc =
+        dnnl::deconvolution_forward::primitive_desc(deconv_desc, attr, engine_);
+
+    // Push to the network.
+    auto deconv = dnnl::deconvolution_forward(deconv2d_prim_desc);
+    net_.push_back(deconv);
+
+    // Data memory.
+    auto deconv2d_src_memory = BindDNNLMemory(data_entry, deconv_src_md);
+
+    // Weight memory.
+    auto deconv2d_weights_memory = BindDNNLMemory(weight_entry, deconv_weights_md);
+
+    // Bias memory.
+    auto deconv2d_bias_memory = dnnl::memory({bias_dims, dt::f32, tag::x}, engine_);
+    if (has_bias) {
+      auto bias_entry = node.GetInputs()[2];
+      BindDNNLMemory(bias_entry, deconv2d_bias_memory);
+    } else {
+      float bias[channels] = {0};
```

Comment:
no need to create and push this arg when `has_bias=false`?

Reply:
no need to create and push. fixed.